fluent/migrate/transforms.py

```diff
@@ -221,59 +276,76 @@ def __init__(self, path, key, replacements):
+    semicolon-separated list of plural forms. Each form will be converted
+    into a TextElement and run through the `foreach` function, which should
+    return an `FTL.Node` or a `Transform`. By default, the `foreach` function
+    creates a valid Pattern from the TextElement passed into it.
     """
     DEFAULT_ORDER = ('zero', 'one', 'two', 'few', 'many', 'other')
 
-    def __init__(self, path, key, selector, foreach=pattern_from_text):
+    def __init__(self, path, key, selector, foreach=Transform.pattern_of):
         super(self.__class__, self).__init__(path, key)
         self.selector = selector
         self.foreach = foreach
 
     def __call__(self, ctx):
-        value = super(self.__class__, self).__call__(ctx)
+        element = super(self.__class__, self).__call__(ctx)
         selector = evaluate(ctx, self.selector)
-        variants = value.split(';')
         keys = ctx.plural_categories
-
-        # A special case for languages with one plural category. We don't need
-        # to insert a SelectExpression at all for them.
-        if len(keys) == len(variants) == 1:
-            variant, = variants
-            return evaluate(ctx, self.foreach(variant))
-
-        # The default CLDR form should be the last we have in
-        # DEFAULT_ORDER, usually `other`, but in some cases `many`.
-        # If we don't have a variant for that, we'll append one,
-        # using the, in CLDR order, last existing variant in the legacy
-        # translation. That may or may not be the last variant.
+        forms = [
+            FTL.TextElement(part)
+            for part in element.value.split(';')
+        ]
+
+        # The default CLDR form should be the last we have in DEFAULT_ORDER,
+        # usually `other`, but in some cases `many`. If we don't have a variant
+        # for that, we'll append one, using the, in CLDR order, last existing
+        # variant in the legacy translation. That may or may not be the last
+        # variant.
         default_key = [
             key for key in reversed(self.DEFAULT_ORDER) if key in keys
         ][0]
 
-        keys_and_variants = zip(keys, variants)
-        keys_and_variants.sort(key=lambda (k, v): self.DEFAULT_ORDER.index(k))
-        last_key, last_variant = keys_and_variants[-1]
+        # Match keys to legacy forms in order they are defined in
+        # Gecko's PluralForm.jsm. Filter out empty forms.
+        pairs = [
+            (key, var)
+            for key, var in zip(keys, forms)
+            if var.value
+        ]
+
+        # A special case for legacy translations which don't define any
+        # plural forms.
+        if len(pairs) == 0:
+            return Transform.pattern_of()
+
+        # A special case for languages with one plural category or one legacy
+        # variant. We don't need to insert a SelectExpression for them.
+        if len(pairs) == 1:
+            _, only_form = pairs[0]
+            return evaluate(ctx, self.foreach(only_form))
```

Comment:
Does this need a `pattern_of`?

Reply:
Technically, it doesn't. My motivation was that with it, all* transforms return via `pattern_of`. It makes the code easier to follow for me and also keeps the text normalization logic in one place.

_* I noticed one code path in `PLURALS` which didn't; I'll fix it._